contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,328 @@
+persistentvolumeclaims. This is needed to avoid situation when compromising node
+de facto means compromising a cluster. For more details & discussions see
+https://github.com/kubernetes/kubernetes/issues/40476.
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual
+objects, by sending individual GET requests. However, it is sending those requests
+periodically to enable automatic updates of mounted secrets/configmaps/...
+In large clusters, this is generating huge unnecessary load, as this load in
+principle should be watch-based. We would like to address this together with
+solving the authorization issue.
+
+
+# Proposal
+
+ In this proposal, we're not focusing on how exactly security should be done.
+We're just sketching very high level approach and exact authorization mechanism
+should be discussed separately.
+
+ At the high level, what we would like to achieve is to enable LIST and WATCH
+requests to support more sophisticated filtering (e.g. we would like to be able
+to somehow ask for all Secrets attached to all pods bound to a given node in a
+single LIST or WATCH request). We obviously want to be able to authorize all
+other types of requests (in particular GETs), so that the design has to be
+consistent in all those cases.
+
+ To solve this problem, we propose to introduce the idea of ```bulk watch ``` and
+```bulk get ``` (and in the future other bulk operations). This idea was already
+appearing in the community and now it seems to be a good time to proceed with it.
+
+ We would like the API to be generic enough so that it will be useful for many
+different usecases, not just for this particular one. As a result, below we are
+describing requirements that the solution has to meet to satisfy our needs:
+- a single bulk request has to support multiple resources (e.g. get a node and
+all pods associated with it)
+- if the bulk request filters asks only for object of a single resource type,
+the returned result should ideally be comparible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are server by
+different apiservers
+
+TODO: What else do we require?
+
+
+# Detailed design
+
+ To make the API generic enough, one of our requirements is to make those bulk
+operations work across different resource types (e.g. watch pod A and secret S
+within a single watch call). [Though, this doesn't need to be supported by the
+initial implementation - this just have to be possible at the API level.]
+
+ In the current API, it is not possible to express a request that is spanning
+multiple different resource types. So as a result, we need to introduce completely
+new API to support it.
```

Comment:
I think this needs a supporting sentence:

Spanning multiple resources, resource types, or conditions will be increasingly important for large numbers of watches.  To ensure we can leave room for future low level operations, we want clients to ask for a large number of watches in a single call.  Since pods can reference multiple resources, we potentially have small M number of resource types, but large N resources or conditions per node.  Nodes currently watch between 2 and 10 resource types, but other integrations like federation will increasingly add more watches for every type they federate.  So bypassing resource type level aggregation and going straight to full aggregation ensures we can more aggressively optimize in the future.

Also, our current REST API does not offer an easy place to handle "multiple watches for this type" - rather than invent a new top level pattern per type, we can simply introduce a new resource type that follows normal RESTful rules.

^ not sure how to exactly word that, but i think it's important to prove we don't want to add this per resource.